A Deeper Look Into Git: turning remote section into lesson (#26118)
TheOdinProject/curriculum: git/intermediate_git/working_with_remotes.md

1.  If working on a team project, make sure rewriting history is safe to do and others know you're doing it.
2.  Ideally, stick to using these commands only on branches that you're working with by yourself.
3.  Using the `-f` flag to force something should scare you, and you better have a really good reason for using it.
4.  Don't push after every single commit, changing published history should be avoided when possible.
5.  Regarding the specific commands we've covered:
    1.  For `git amend` never amend commits that have been pushed to remote repositories.
    2.  For `git rebase` never rebase a repository that others may work off of.
    3.  For `git reset` never reset commits that have been pushed to remote repositories.
    4.  For `git push --force` only use it when appropriate, use it with caution, and preferably default to using `git push --force-with-lease`.

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Read the chapter on [Branches covered by git-scm](https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell) if you want an even deeper dive into branches.
2. Read [think-like-a-git](http://think-like-a-git.net/)

>     *   Take your time with this resource as well, it's very well written and will be very helpful in solidifying your understanding of Git.

Reviewer: I would keep the merge conflicts reading here.

Author: Made the changes @wise-king-sullyman.

Reviewer: Do you not want to update the Deeper Look at Git lesson as part of this PR?

Author: Oh ok I'll gladly do it @wise-king-sullyman

Author: Made the change @wise-king-sullyman

Reviewer: It looks like you removed the entire bottom half of the Deeper Look at Git lesson? We want to keep the branches are pointers section, the relevant assignments, knowledge checks, and additional resources for that lesson.

Also you should update the lesson overview points.

Author: Ok @wise-king-sullyman. I'll add those back in. Can you point out which parts I should remove?

Reviewer: All that should be removed from the Deeper Look at Git lesson is the stuff you added to the new remotes lesson.

Author: Ok

</div>

### Knowledge check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

- [A KNOWLEDGE CHECK QUESTION](A-KNOWLEDGE-CHECK-URL)

### Additional resources

This section contains helpful links to related content. It isn’t required, so consider it supplemental.

> * If you're looking for a fun way to deepen your knowledge of working with branches and Git, check out the popular game, [Learn Git Branching](https://learngitbranching.js.org/)

Reviewer: ```suggestion
* If you're looking for an interactive way to deepen your knowledge of working with Git, check out this game, [Learn Git Branching](https://learngitbranching.js.org/)
```
We don't really know how popular it is, and while I think it's a good tool I don't think many would find it "fun" in the typical sense.

Also the "branches and" seems unneeded since branches are part of Git.

The `the` -> `this` change just makes it read more naturally to me, but if that's grammatically incorrect just ignore this part, I'm no subject matter expert on grammar.

So now that we've learned about the various dangers of `git push --force`, you're probably wondering why it exists and when to use it. A very common scenario in which developers use `git push --force` is updating pull requests. Collaborative work is covered more in depth in a separate lesson, but the take-away from this section should be that the `--force` option should be used only when you are certain that it is appropriate. There are also less common scenarios, such as when sensitive information is accidentally uploaded to a repository and you want to remove all occurrences of it.

<span id='force-with-lease'>It is worth giving special mention to `git push --force-with-lease`</span>, a command which in some companies is the default option. The reason for this is that it's a fail-safe! It checks if the branch you're attempting to push to has been updated and sends you an error if it has. This gives you an opportunity to, as mentioned before, `fetch` the work and update your local repository.

> ### Dangers and best practices

Reviewer: You'll need a newline here for[ formatting reasons](https://www.theodinproject.com/lessons/preview?uuid=c0126432-aac9-412d-8c1e-79d48d6ad3e0) as well.

> ### Dangers and best practices

Reviewer: ```suggestion

### Dangers and best practices
```

1.  Read through [GitHub's documentation on merge conflicts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/about-merge-conflicts)
    *   It's only a matter of time until you run into one (if you haven't already)! While merge conflicts might seem intimidating, they're actually very simple. Take your time with this resource and make sure you look at the two different ways the documentation suggests resolving merge conflicts - on GitHub itself, and on your command line. While you might not need this right now, keeping the source of this documentation in the back of your mind will prove invaluable for when you eventually run into a merge conflict and aren't sure where to find a simple solution.

2.  Read [think-like-a-git](http://think-like-a-git.net/)
    *   Take your time with this resource as well, it's very well written and will be very helpful in solidifying your understanding of Git.
</div>

### Knowledge check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

> *   <a class='knowledge-check-link' href='https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History'>What are some different ways to rewrite history?</a>

Reviewer: I feel like this knowledge check goes better with the deeper look at Git lesson.

*   <a class='knowledge-check-link' href='#dangers'>What are the dangers of history-changing operations?</a>
*   <a class='knowledge-check-link' href='#best-practices'>What are best practices of history-changing operations?</a>
*  <a class='knowledge-check-link' href='https://git-scm.com/book/en/v2/Git-Basics-Undoing-Things'>How can you amend your last commit?</a>

> 

Reviewer: The "amend your last commit" lesson goes into the previous lesson.
